### PR TITLE
Cascading: java flight server + relay

### LINKED ISSUE #16: Benchmarking of cascaded flight servers
@cdoron to edit with plan details

### PR DESCRIPTION
**Pull request by cdoron**
initial version
resolves #16

Signed-off-by: Doron Chen <cdoron@il.ibm.com>

### CODE REVIEW

File: Java-FlightServer/src/flight-client/main/java/flight_client/org/m4d/adp/flight_client/RunBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.apache.arrow.flight.*;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.util.AutoCloseables;
+import org.apache.arrow.vector.VectorSchemaRoot;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+public class RunBenchmark {
```
Comment (simanadler) [inline Java-FlightServer/src/flight-client/main/java/flight_client/org/m4d/adp/flight_client/RunBenchmark.java]: Please add comments describing what it is you are testing and how you are testing it.  In general some documentation for all the functions would be helpful

File: Java-FlightServer/src/flight-client/main/java/flight_client/org/m4d/adp/flight_client/MyFlightClient.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
+package org.m4d.adp.flight_client;
+
+import org.apache.arrow.flight.*;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.VectorSchemaRoot;
+
+import java.nio.charset.StandardCharsets;
+
+public class MyFlightClient {
```
Comment (simanadler) [inline Java-FlightServer/src/flight-client/main/java/flight_client/org/m4d/adp/flight_client/MyFlightClient.java]: Please add some comments describing what this client does.

File: Java-FlightServer/src/flight-server/main/java/org/m4d/adp/flight_server/ExampleProducer.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.concurrent.Executors;
+
+public class ExampleProducer extends NoOpFlightProducer {
+    private final Location location;
+    private final BufferAllocator allocator;
+    private final int RecordsPerBatch = 1024*1024;
+    private final VectorSchemaRoot constVectorSchemaRoot;
+    private boolean isNonBlocking = false;
+
+    public ExampleProducer(Location location, BufferAllocator allocator) {
```
Comment (simanadler) [inline Java-FlightServer/src/flight-server/main/java/org/m4d/adp/flight_server/ExampleProducer.java]: Can you please add documentation here and throughout this file?

File: Java-FlightServer/src/flight-server/main/java/org/m4d/adp/flight_server/RelayProducer.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.arrow.vector.VectorUnloader;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class RelayProducer extends NoOpFlightProducer {
```
Comment (simanadler) [inline Java-FlightServer/src/flight-server/main/java/org/m4d/adp/flight_server/RelayProducer.java]: Again, please document

**Review by simanadler — COMMENTED** [review]
Thanks for this initial PR @cdoron! Please add to the PR a link to the issue and/or document describing the problem or feature you are addressing.  You mention that this is an initial version.  Please indicate what parts have and have not been addressed by this PR. Code needs documentation as well.    FYI - I did not review the code in depth.

**Comment by the-mesh-for-data** [comment]
Addresses Issue #16

**Comment by cdoron** [comment]
addresses issue #16

**Review by roee88 — APPROVED** [review]

**Comment by the-mesh-for-data** [comment]
Hi @simanadler : I added comments in a separate PR.
This PR contains the code for a flight server which contains a single dataset in memory, a relay flight server which gets a record-batch and transmits a record batch (either identical or after a transformation) and client code.
I labeled the PR 'initial' since there is still further investigations to be done on how to improve performance of the relay server. Also, we currently support only a single hard-coded transformation. We still need to determine the proper design for supporting multiple transformations.

**Comment by cdoron** [comment]
Hi @simanadler : I added comments in a separate PR.
This PR contains the code for a flight server which contains a single dataset in memory, a relay flight server which gets a record-batch and transmits a record batch (either identical or after a transformation) and client code.
I labeled the PR 'initial' since there is still further investigations to be done on how to improve performance of the relay server. Also, we currently support only a single hard-coded transformation. We still need to determine the proper design for supporting multiple transformations.

**Pull request merged**